src/swap/v2/router/swap-router.ts

```diff
@@ -0,0 +1,198 @@
+    appArgs: [V2_SWAP_ROUTER_APP_ARGS_ENCODED.ASSET_OPT_IN],
+    foreignAssets: assetIDs,
+    suggestedParams
+  });
+  // The number of inner transactions is the number of assets we're opting in to
+  const innerTransactionCount = assetIDs.length;
+
+  // The fee for the transaction is the fee for the outer transaction (which is a NoOp)
+  // multiplied by the number of inner transactions
+  assetOptInTxn.fee = ALGORAND_MIN_TX_FEE * (innerTransactionCount + 1);
+
+  return assetOptInTxn;
+}
+
+export async function generateSwapRouterTxns({
+  initiatorAddr,
+  client,
+  network,
+  swapType,
+  route
+}: GenerateSwapRouterTxnsParams) {
+  const suggestedParams = await client.getTransactionParams().do();
+
+  const [assetInID, intermediaryAssetID, assetOutID] = [
+    getAssetId(route[0].quote.amount_in.asset),
+    getAssetId(route[0].quote.amount_out.asset),
+    getAssetId(route[1].quote.amount_out.asset)
+  ];
+  const [assetInAmount, assetOutAmount] = [
+    Number(route[0].quote.amount_in.amount),
+    Number(route[route.length - 1].quote.amount_out.amount)
+  ];
+  const [pool1Address, pool2Address] = [route[0].pool.address, route[1].pool.address];
+
+  const isAssetInAlgo = isAlgo(assetInID);
+
+  const routerAppID = getSwapRouterAppID(network);
+
+  const inputTxn = isAssetInAlgo
+    ? algosdk.makePaymentTxnWithSuggestedParamsFromObject({
+        from: initiatorAddr,
+        to: getApplicationAddress(routerAppID),
+        amount: assetInAmount,
+        suggestedParams
+      })
+    : algosdk.makeAssetTransferTxnWithSuggestedParamsFromObject({
+        from: initiatorAddr,
+        to: getApplicationAddress(routerAppID),
+        amount: assetInAmount,
+        assetIndex: assetInID,
+        suggestedParams
+      });
+
+  const routerAppCallTxn = algosdk.makeApplicationNoOpTxnFromObject({
+    from: initiatorAddr,
+    appIndex: routerAppID,
+    appArgs: [
+      V2_SWAP_APP_CALL_ARG_ENCODED,
+      V2_SWAP_APP_CALL_SWAP_TYPE_ARGS_ENCODED[swapType],
+      algosdk.encodeUint64(assetOutAmount)
+    ],
+    foreignApps: [getValidatorAppID(network, CONTRACT_VERSION.V2)],
+    foreignAssets: [assetInID, intermediaryAssetID, assetOutID],
+    accounts: [pool1Address, pool2Address],
+    suggestedParams
+  });
+
+  const innerTransactionCount = swapType === SwapType.FixedInput ? 7 : 8;
+
+  routerAppCallTxn.fee = ALGORAND_MIN_TX_FEE * (innerTransactionCount + 1);
+
+  const txnList = [inputTxn, routerAppCallTxn];
+
+  const swapRouterAppOptInRequiredAssetIDs = await getSwapRouterAppOptInRequiredAssetIDs({
+    client,
+    routerAppID,
+    assetIDs: [assetInID, intermediaryAssetID, assetOutID]
+  });
+
+  if (swapRouterAppOptInRequiredAssetIDs.length > 0) {
+    const routerAppOptInTxn = await generateSwapRouterAssetOptInTransaction({
+      client,
+      accountAddress: initiatorAddr,
+      assetIDs: swapRouterAppOptInRequiredAssetIDs,
+      routerAppID
+    });
+
+    txnList.unshift(routerAppOptInTxn);
```

Comment:
```suggestion

    // Asset opt-in transaction should be the first one in the group
    txnList.unshift(assetOptInTxn);
```
We are calling it "routerAppOptInTxn", but it's actually "asset opt-in transaction, no? Maybe we can imporove the naming.
Also;
`swapRouterAppOptInRequiredAssetIDs` -> `optInRequiredAssetIDs`